src/peergos/shared/user/SharedWithCache.java

```diff
@@ -9,59 +12,75 @@
 
     public enum Access { READ, WRITE }
 
-    private Map<String, Set<String>> sharedWithReadAccessCache = new ConcurrentHashMap<>(); //path to friends
-    private Map<String, Set<String>> sharedWithWriteAccessCache = new ConcurrentHashMap<>();
+    //K,V
+    // K = mapKey of absolute capability
+    // V = list of usernames the file/dir is shared with
+    private Map<ByteArrayWrapper, Set<String>> sharedWithReadAccessCache = new ConcurrentHashMap<>();
+    private Map<ByteArrayWrapper, Set<String>> sharedWithWriteAccessCache = new ConcurrentHashMap<>();
 
     public SharedWithCache() {
 
     }
 
-    private String canonicalisePath(String path) {
-        return path.startsWith("/") ? path : "/" + path;
+    private ByteArrayWrapper generateKey(AbsoluteCapability cap) {
+        return new ByteArrayWrapper(cap.getMapKey());
+    }
+
+    public boolean isShared(AbsoluteCapability cap) {
+        return ! getSharedWith(SharedWithCache.Access.READ, cap).isEmpty()
+                || ! getSharedWith(SharedWithCache.Access.WRITE, cap).isEmpty();
+    }
+
+    public void copySharedWith(AbsoluteCapability oldCap, AbsoluteCapability newCap) {
+        Set<String> sharedReadAccessWith = getSharedWith(SharedWithCache.Access.READ, oldCap);
+        Set<String> sharedWriteAccessWith = getSharedWith(SharedWithCache.Access.WRITE, oldCap);
+        addSharedWith(SharedWithCache.Access.READ, newCap, sharedReadAccessWith);
+        addSharedWith(SharedWithCache.Access.WRITE, newCap, sharedWriteAccessWith);
     }
 
-    public Set<String> getSharedWith(Access access, String path) {
+    public Set<String> getSharedWith(Access access, AbsoluteCapability cap) {
         return access == Access.READ ?
-            getSharedWith(sharedWithReadAccessCache, path) : getSharedWith(sharedWithWriteAccessCache, path);
+            getSharedWith(sharedWithReadAccessCache, cap) : getSharedWith(sharedWithWriteAccessCache, cap);
     }
 
-    private synchronized Set<String> getSharedWith(Map<String, Set<String>> cache, String path) {
-        return new HashSet<>(cache.computeIfAbsent(path, k -> new HashSet<>()));
+    private synchronized Set<String> getSharedWith(Map<ByteArrayWrapper, Set<String>> cache, AbsoluteCapability cap) {
+        return new HashSet<>(cache.computeIfAbsent(generateKey(cap), k -> new HashSet<>()));
```

Comment:
probably don't want to keep filling this map up as we browse around. could just use getOrDefault?